integration/smartbft/smartbft_block_deliverer_test.go

```diff
@@ -0,0 +1,205 @@
+	closeTxProducer := make(chan bool)
+	go func() {
+		defer GinkgoRecover()
+		var i uint64
+		for {
+			buff := make([]byte, 8)
+			binary.BigEndian.PutUint64(buff, i)
+
+			/* Create transaction envelope which will be broadcast to all orderers */
+			env := ordererclient.CreateBroadcastEndorserTransactionEnvelope(network, network.Orderers[1], channel, buff)
+
+			for _, stream := range allStreams {
+				err := stream.Send(env)
+				Expect(err).NotTo(HaveOccurred())
+			}
+			i++
+
+			finishSending := false
+			select {
+			case <-closeTxProducer:
+				finishSending = true
+			case <-time.After(time.Second):
+			}
+			if finishSending {
+				return
+			}
+		}
+	}()
+	Eventually(ordererRunners[1].Err(), 10*time.Second).Should(gbytes.Say("Delivering proposal, writing block 10"))
+	closeTxProducer <- true
+
+	/* Shut down the orderers */
+
+	By("Taking down all the orderers")
+	for _, proc := range ordererProcesses {
+		proc.Signal(syscall.SIGINT)
+		Eventually(proc.Wait(), network.EventuallyTimeout).Should(Receive())
+	}
+
+	/* Extract ledger from one of the orderers */
+	By("Extracting blocks from ledger")
+	ordererFabricConfig := network.ReadOrdererConfig(network.Orderers[0])
+	blockStoreProvider, err := blkstorage.NewProvider(
+		blkstorage.NewConf(ordererFabricConfig.FileLedger.Location, -1),
+		&blkstorage.IndexConfig{
+			AttrsToIndex: []blkstorage.IndexableAttr{blkstorage.IndexableAttrBlockNum},
+		},
+		&disabled.Provider{},
+	)
+	Expect(err).NotTo(HaveOccurred())
+	blockStore, err := blockStoreProvider.Open(channel)
+	Expect(err).NotTo(HaveOccurred())
+	fileLedger := fileledger.NewFileLedger(blockStore)
+	var ledgerArray []*common.Block
+	for i := uint64(0); i < fileLedger.Height(); i++ {
+		block, err := fileLedger.RetrieveBlockByNumber(i)
+		Expect(err).NotTo(HaveOccurred())
+		ledgerArray = append(ledgerArray, block)
+	}
+	/* The ledger now saved and can be accessed at ledgerArray */
+
+	/* Create orderer mocks and make sure they can access the ledger */
+	var mocksArray [4]*MockOrderer
+
+	for i, orderer := range network.Orderers {
+		serverTLSCerts := map[string][]byte{}
+		tlsCertPath := filepath.Join(network.OrdererLocalTLSDir(orderer), "server.crt")
+		serverTLSCerts[tlsCertPath], err = os.ReadFile(tlsCertPath)
+		Expect(err).NotTo(HaveOccurred())
+
+		serverTLSKeys := map[string][]byte{}
+		tlsKeyPath := filepath.Join(network.OrdererLocalTLSDir(orderer), "server.key")
+		serverTLSKeys[tlsKeyPath], err = os.ReadFile(tlsKeyPath)
+		Expect(err).NotTo(HaveOccurred())
+
+		secOpts := comm.SecureOptions{
+			UseTLS:      true,
+			Certificate: serverTLSCerts[tlsCertPath],
+			Key:         serverTLSKeys[tlsKeyPath],
+		}
+		mo, err := NewMockOrderer(network.OrdererAddress(orderer, nwo.ListenPort), ledgerArray, secOpts)
+		Expect(err).NotTo(HaveOccurred())
+		mo.logger.Infof("Mock orderer started at port %v", network.OrdererAddress(orderer, nwo.ListenPort))
+		mocksArray[i] = mo
+	}
+
+	/* Create peer */
```

Comment:
up until here it is the preparation of the mocks with ledgers. I think you can extract most of it to utility functions call this function in a "Before". 

From here on, it's the actual tests. Currently you have a mock that behaves like a correct orderer. This is test number 1 and it works fine. It should be in an `It("correct mock")`

In additional tests we will change the mock to exhibit byzantine behavior, each type of behavior in a separate "It". Let do that in separate PRs, lets finish this basic one first.

Reply:
Done